### ISSUE TITLE
Differential activity analysis of positive and negative reactions

### ISSUE THREAD

**Issue by Hrovatin** [YosefLab/Compass#47]
I am trying to analyse the results of compass. However, for reversible reactions (having both pos and neg) it is not clear to me which one I should use. Namely, often both directions would lead to large differential activity which would be in the same direction for both of them (but I would expect the opposite, namely if pos is upregulated then neg should be downregulated if both are diferential). I assume that I can not just pick one direction with higher consistency across cells because some cells could have positive and the others negative reaction active.
At the end I would like to get one effect size score per reaction so that I can plot them in Recon Map.

**Comment by schel337** [comment]
That's interesting, I would also tend to think that there would not generally be a large differentia for both directions of a reaction.  I'm going to defer to Allon on how to interpret this, and I've pinged him on it.

**Comment by allonw** [comment]
That's a good question. The two directions can be significantly active in different directions (and then you've found an interesting metabolic control point!), but frequently, both directions of the reactions will be differential in the same directions, as you described. This occurs because of the presence of cycles in the network.

If your purpose is visualization, then the intuitive solution would be to draw one directed edge per reaction, based for example on the direction shown in KEGG maps, and pick the score corresponding to the chosen direction

**Comment by Hrovatin** [comment]
There are many such reactions that seem to have cycles (e.g. both positive and negative reaction are upregulated in one of the groups). My cells come from different studies and the Compass input data was jointly normalised, but not integrated on expression level. Could this lead to some kind of biases that result in reaction scores being in generally higher in one group of cells that comes mainly from a different set of studies?
Here are my Cliff's delta values distributions for reversible reactions:
![image](https://user-images.githubusercontent.com/47607471/133731450-09caebea-e424-4f09-a57a-72ffdaccc206.png)

I have tried mapping Recon3 to KEGG before, but we so far managed to map only ~10% of Recon reactions. Do you have a better mapping?

**Comment by Hrovatin** [comment]
Indeed, the reason for more upregulated genes than downregulated genes seems to be more metabolic counts in one of the groups (e.g. young (case) compared to adult (control)).

Metabolic gene counts distn (the t2d sample is not used here):
![image](https://user-images.githubusercontent.com/47607471/133886392-ff58ce9c-4b45-40e9-b51f-acda21f60862.png)

All genes counts distn:
![image](https://user-images.githubusercontent.com/47607471/133886406-d85a69e8-9e11-4c42-bb22-7efa2f912e91.png)

Would you suggest doing normalisation on metabolic genes only?

**Comment by allonw** [comment]
> I have tried mapping Recon3 to KEGG before, but we so far managed to map only ~10% of Recon reactions. Do you have a better mapping?

How do you map them? With the KEGG ID given as Recon's metadata or by name? In many cases the Recon annotation is incomplete because of computational flukes, and reactions that have a KEGG parallel aren't annotated with the correct KEGG ID and can be matched only based on EC or associated gene.

> There are many such reactions that seem to have cycles (e.g. both positive and negative reaction are upregulated in one of the groups). If this are just some kind of futile cycles then I assume this still represents reaction activity and I could visualise just the maximum effect size of the concerned reactions?

Yes, the metabolic network is so highly connected that many reactions are on futile cycles. In reality, most (but not all) of these cycles are inoperative because of the second law of thermodynamics, but this can't be tractably modeled in the FBA framework.

I would still prefer a heuristic that consistently directs the edges in some way (e.g., from source to end-product) and allows you to pick the compass score of the right direction, but your idea can work as long as one remembers that scores of adjacent compass reaction could come from flux distributions in which the reaction in question carried flux in opposite directions.

**Comment by allonw** [comment]
> Indeed, the reason for more upregulated genes than downregulated genes seems to be more metabolic counts in one of the groups (e.g. young (case) compared to adult (control)).

Yes, this is very common. The baseline activity metabolic activity level in independent experiments can be different, and bias all test results in one direction. When we had similar situations, we normalized only on metabolic genes - exactly as you suggested.

I'll just add that the difference can also be biological and not technical. For example, it can be that cell type A has overall higher metabolic activity (e.g., it's proliferating) than cell type B. In that case, it makes sense to look both at this absolute difference, and having removed it to test for differential metabolic rewiring between the two cell types. For example, A may carry more flux in absolute terms on reaction X, but B may carry more flux on X relative to its overall metabolic activity.

In the Compass [paper](https://doi.org/10.1016/j.cell.2021.05.045), Fig. S2A, we show that the overall metabolic activity corresponds to PC1 of our dataset, and that the subsequent PCs reflect different metabolic usages.

In this sense, metabolic activity of a cell is a covariate that is not purely technical nuisance and not purely biological signal, which is more common in single-cell datasets than it was in bulk genomics, as we discussed [here](https://doi.org/10.1038/nbt.3711).

**Comment by Hrovatin** [comment]
Thank you for the extensive response.

> How do you map them? With the KEGG ID given as Recon's metadata or by name? In many cases the Recon annotation is incomplete because of computational flukes, and reactions that have a KEGG parallel aren't annotated with the correct KEGG ID and can be matched only based on EC or associated gene.

We used combination of KEGG id mappings in Recon metadata and info retrieved from MetRxn and BiKEGG. This leads to quite ok coverage of many human KEGG pathways (many >50%), but Recon coverage is around 10%. Also, some of the KEGG reactions we mapped in this way are not contained in human-associated KEGG pathways.
So you suggest also mapping reactions based on EC or genes? How do you deal with cases where same Ec catalyses multiple reactions with different substrates/products?

> I'll just add that the difference can also be biological and not technical. For example, it can be that cell type A has overall higher metabolic activity (e.g., it's proliferating) than cell type B. In that case, it makes sense to look both at this absolute difference, and having removed it to test for differential metabolic rewiring between the two cell types. For example, A may carry more flux in absolute terms on reaction X, but B may carry more flux on X relative to its overall metabolic activity.

So would it be better to start from metabolic-genes specific normalisation or to normalise reaction activities obtained from Compass? I am not sure if the latter would be a valid approach or not (e.g. how may it affect Compass computation itself if we used different expression normalisations, especially with information sharing across cells). 

>In this sense, metabolic activity of a cell is a covariate that is not purely technical nuisance and not purely biological signal, which is more common in single-cell datasets than it was in bulk genomics, as we discussed here.

I guess it would be hard to determine which one it really is based on transcriptomic data alone; or do you have any suggestion? If we do not know what is the cause of seemingly changed metabolic activity then I agree that relative comparisons are the safest option.
